### PR TITLE
Support NDP Router Advertisement + source link layer options

### PR DESCRIPTION
**Pull request by scrollins**
Support router advertisement + source link layer options

Will probably need some refactoring as you have to manually set the checksum + set the length. I was able to set the checksum by loading the pcap into wireshark and copying what the expected checksum was supposed to be. Kind of a hack. Also, you have to set the packet length manually. Options can be a variable size but all packet lengths will be 70 since we only support one type of option right now. Once we go to supporting multiple options, we might want to refactor and have it set the packet length like ip6sr does. Things to also consider. Perhaps we move the options parsing to its own module since it would be the same across multiple NDPs. Router advertisement message is the only NDP message we support now. 

A sample input would look like the following
```
return {
  {
    id = "icmp6RouterAdvertisement",
    args = {packetLength = 78},
    update = function(buf)
      local pkt = buf:getIcmp6RouterAdvertisement()
      pkt.eth:setSrcString("00:00:00:00:00:00")
      pkt.eth:setDstString("00:00:00:00:00:01")
      pkt.ip6:setSrcString("c1fe:4700:add2::1111")
      pkt.ip6:setDstString("5e2f:1ce0:add2::")
      pkt.icmp6RouterAdvertisement:setReserved(0);
      pkt.icmp6RouterAdvertisement:setRetransTimer(111);
      pkt.icmp6RouterAdvertisement:setReachableTime(10);
      pkt.icmp6RouterAdvertisement:setType(134);
      pkt.icmp6RouterAdvertisement:setChecksum(0x7078);
      pkt.icmp6RouterAdvertisement:setSourceLinkLayerOptions("c2:00:54:f5:00:00");
    end
  },
}

### CODE REVIEW

File: lua/proto/ip6/icmp/router_advertisement.lua [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    self.retrans_timer = bswap(int)
+end
+
+function icmp6RouterAdvertisementHeader:setReachableTime(int)
+    int = int or 0
+    self.reachable_time = bswap(int)
+end
+
+function icmp6RouterAdvertisementHeader:setOptions(options)
+    for k,v in pairs(options) do
```
Comment (drunkirishcoder) [inline lua/proto/ip6/icmp/router_advertisement.lua]: since this is the only option we can support. maybe just not do the loop?

**Comment by scrollins** [comment]
i went ahead and removed icmp6 module since that was copied over. it wasn't being used by anything. i created it to try to keep the same framework but realized icmp6 wouldnt work. that's probably why the libmoon maintainers chose to punt on trying to implement icmp6. its really hard to do w/ the different formats  + the lack of options structures defined in the RFC. i dont love the fact that you would have to dup code for other message types if we choose to use libmoon for implementing other ndp messages

File: lua/proto/ip6/icmp/router_advertisement.lua (hunk 1 above)
Comment (scrollins) [reply lua/proto/ip6/icmp/router_advertisement.lua]: i think prefix information we will need to support in the near future so i left this as a loop

File: lua/proto/ip6/icmp/router_advertisement.lua [hunk 2]
```diff
@@ -0,0 +1,96 @@
+
+local icmp6RouterAdvertisement = {}
```
[the remark this answers is not in the document]
Comment (scrollins) [reply lua/proto/ip6/icmp/router_advertisement.lua]: if there is a way to do this using subclassing, i'm open to it so we dont have to dup code

File: lua/proto/ip6/icmp/router_advertisement.lua (hunk 1 above)
Comment (drunkirishcoder) [reply lua/proto/ip6/icmp/router_advertisement.lua]: as we discussed, libmoon can't really be dynamic. a router advertisement with just the source link layer address has to be a different type from one with both source link layer address and prefix info. this loop wouldn't support what we want anyway. might as well just lay it out as fixed fields at the end.

File: lua/packet.lua [hunk 3]
```diff
@@ -884,7 +885,19 @@ end
 pkt.getArpPacket = createStack("eth", "arp")
 
 pkt.getIcmp4Packet = createStack("eth", "ip4", "icmp")
-pkt.getIcmp6Packet = createStack("eth", "ip6", "icmp")
+pkt.getIcmp6Packet = createStack("eth", "ip6", "icmp6")
+pkt.getIcmp6RouterAdvertisement = createStack("eth", "ip6", "icmp6RouterAdvertisement")
+
+pkt.getTcpPacket = function(self, ip4)
```
[the remark this answers is not in the document]
Comment (scrollins) [reply lua/packet.lua]: removed

File: lua/proto/constants.lua [hunk 4]
```diff
@@ -14,5 +14,7 @@ constants.PROTO_ESP  = 0x32 -- 50
 constants.PROTO_AH   = 0x33 -- 51
 --- NextHeader field value for Icmp for IPv6
 constants.PROTO_ICMP = 0x3a -- 58
+--- NextHeader field value for Icmp for IPv6
+constants.PROTO_ICMP6 = 0x3a -- 58
```
[the remark this answers is not in the document]
Comment (clinedome) [reply lua/proto/constants.lua]: looks like the other one was for ICMPv6 anyway

File: lua/packet.lua (hunk 3 above)
[the remark this answers is not in the document]
Comment (clinedome) [reply lua/packet.lua]: is it?

File: lua/proto/ip6/icmp/router_advertisement.lua (hunk 2 above)
[the remark this answers is not in the document]
Comment (clinedome) [reply lua/proto/ip6/icmp/router_advertisement.lua]: yeah, i do feel like there might be a way, but i'd have to tinker with it. all told, if we could get the subclassing working, I think that hard-coding the different variations of these messages that we actually care about is actually completely reasonable. we don't really need this to support the entire breadth of what icmpv6 offers. it certainly would be possible, if ugly, to do that -- but i don't think the benefit is likely worth the effort.

File: lua/proto/ip6/icmp/router_advertisement.lua [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+icmp6RouterAdvertisementHeader.__index = icmp6RouterAdvertisementHeader
+
+-- definition of the header format
+icmp6RouterAdvertisement.headerFormat = [[
+	struct      icmp6;
+	uint16_t    reserved;
+	uint16_t	router_lifetime;
+	uint32_t	reachable_time;
+	uint32_t	retrans_timer;
+	union       source_link_layer_address_option source_link_layer;
```
Comment (drunkirishcoder) [inline lua/proto/ip6/icmp/router_advertisement.lua]: some mixed spaces and tabs actions here.

File: lua/proto/ip6/icmp/icmp6.lua [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+require "proto.template"
+require "proto.ethernet"
+local ffi = require "ffi"
+
+-- struct for ndp
+ffi.cdef[[
+	struct icmp6 {
+	uint8_t     type;
+	uint8_t     code;
+	uint16_t    checksum;
```
Comment (drunkirishcoder) [inline lua/proto/ip6/icmp/icmp6.lua]: could use some indentation here and a few other places to make this a little more readable. and also there are tabs.

File: lua/proto/ip6/icmp/router_advertisement.lua [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    self.source_link_layer.uint8[3] = bit.band(bit.rshift(mac_address, 8), 0xFF)
+    self.source_link_layer.uint8[4] = bit.band(bit.rshift(mac_address, 16), 0xFF)
+    self.source_link_layer.uint8[5] = bit.band(bit.rshift(mac_address, 24), 0xFF)
+    self.source_link_layer.uint8[6] = bit.band(bit.rshift(mac_address + 0ULL, 32ULL), 0xFF)
+    self.source_link_layer.uint8[7] = bit.band(bit.rshift(mac_address + 0ULL, 40ULL), 0xFF)
+
+end
+
+icmp6RouterAdvertisement.metatype = icmp6RouterAdvertisementHeader
+return icmp6RouterAdvertisement
```
Comment (drunkirishcoder) [inline lua/proto/ip6/icmp/router_advertisement.lua]: also add a new line here for eof.

File: lua/packet.lua [hunk 8]
```diff
@@ -884,7 +885,18 @@ end
...
+	ip4 = ip4 == nil or ip4
+	if ip4 then
+		return pkt.getTcp4Packet(self)
+	else
+		return pkt.getTcp6Packet(self)
+	end
+end
```
[the remark this answers is not in the document]
Comment (scrollins) [reply lua/packet.lua]: removed

**Pull request merged**